fix: pricing rule item code UOM apply & conversions

maharshivpatel opened the pull request:
There were multiple issues with pricing rule's apply_on_field ==  "item_code" ( UOM : blank, other_than_stock_uom etc ). I have added the required checks and proper conversions.

### Before:
https://user-images.githubusercontent.com/39730881/195068408-1a3a2d4a-7aca-4b3e-8c54-397dd471ba68.mov

### After:
https://user-images.githubusercontent.com/39730881/195422233-1283adb7-447b-441b-8fbd-2a3ab8d74caa.mov

maharshivpatel committed:
fix: pricing rule item_code UOM apply & conversions

there were multiple issues with pricing rule's apply_on_field ==  "item_code" ( UOM : blank, other_than_stock_uom etc ). I have added required checks and proper conversions.

maharshivpatel committed:
Merge branch 'develop' into fix-pricing-rule-uom

deepeshgarg007 committed:
Merge branch 'develop' into fix-pricing-rule-uom

deepeshgarg007 commented on erpnext/accounts/doctype/pricing_rule/pricing_rule.py:
```diff
@@ -440,14 +442,19 @@ def apply_price_discount_rule(pricing_rule, item_details, args):
 			# if pricing_rule_rate > 0
-			item_details.update(
-				{
-					"price_list_rate": pricing_rule_rate * args.get("conversion_factor", 1),
```
So the conversion factor coming from the args should be the `rule_conversion_factor` itself. Why this is not coming investigating and fixing that would be a proper fix

deepeshgarg007 commented on erpnext/accounts/doctype/pricing_rule/utils.py:
```diff
@@ -111,6 +111,12 @@ def _get_pricing_rules(apply_on, args, values):
 		)
 
 		if apply_on_field == "item_code":
+			if args.get("uom") != args.get("stock_uom"):
+				item_conditions += (
+					" and ({child_doc}.uom='{item_uom}' or IFNULL({child_doc}.uom, '')='')".format(
+						child_doc=child_doc, item_uom=args.get("uom")
+					)
+				)
```
What this exactly does I am not able to understand, please explain this as well

deepeshgarg007 reviewed (changes requested):
Also need a test case for this, please add that too

maharshivpatel committed:
fix: pricing rule uom blank or matches sales_uom & tests

Instead of handling multiple cases / conversion i have simplified the rule application.

if pricing rule uom is blank apply on all UOMs and multiple by conversion_factor.

if pricing rule uom is specified apply rule only if sales_uom matches the pricing rule uom.

also added test for both cases.

maharshivpatel committed:
Merge branch 'develop' into fix-pricing-rule-uom

maharshivpatel committed:
chore: remove unused import

remove unused import

maharshivpatel committed:
fix: pricing rule uom on server side

pricing rule when called again on_save used get_cached_doc for performance reasons and didn't have uom parameter so looped in child table and specified matching uom.

maharshivpatel committed:
Merge branch 'develop' into fix-pricing-rule-uom

deepeshgarg007 committed:
Merge branch 'develop' into fix-pricing-rule-uom

deepeshgarg007 committed:
Merge branch 'develop' into fix-pricing-rule-uom

deepeshgarg007 commented:
@Mergifyio backport version-13-hotfix version-14-hotfix

sjain17 commented:
Thank you for fixing this bug; however, it is still not considering the max qty limitations.
